service/datasync/exec_data_sync.go

```diff
@@ -0,0 +1,194 @@
+
+	cfg := DefaultConfig
+	for _, option := range options {
+		option(&cfg)
+	}
+
+	decOptions := cbor.DecOptions{
+		ExtraReturnErrors: cbor.ExtraDecErrorUnknownField,
+		MaxArrayElements:  math.MaxUint32,
+	}
+	decoder, err := decOptions.DecMode()
+	if err != nil {
+		panic(err)
+	}
+
+	var exeDataApi execData.ExecutionDataAPIClient
+	if client == nil {
+		exeDataApi = getAPIClient(execDataAddr)
+	} else {
+		exeDataApi = client
+	}
+
+	e := ExecDataSync{
+		log:         log.With().Str("component", "exec_data_sync").Logger(),
+		decoder:     decoder,
+		execDataApi: exeDataApi,
+		queue:       archive.NewDeque(),
+		buffer:      archive.NewDeque(),
+		limit:       cfg.BufferSize,
+		busy:        0,
+	}
+
+	for _, blockID := range cfg.CatchupBlocks {
+		e.queue.PushFront(blockID)
+		e.log.Debug().Hex("block", blockID[:]).Msg("execution record queued for catch-up")
+	}
+
+	return &e
+}
+
+func getAPIClient(addr string) execData.ExecutionDataAPIClient {
+	// connect to Archive-Access instance
+	MaxGRPCMessageSize := 1024 * 1024 * 20 // 20MB
+	conn, err := grpc.Dial(addr, grpc.WithDefaultCallOptions(grpc.MaxCallRecvMsgSize(MaxGRPCMessageSize)))
+	if err != nil {
+		panic(fmt.Sprintf("unable to create connection to node: %s", addr))
+	}
+	return execData.NewExecutionDataAPIClient(conn)
+}
+
+// OnBlockFinalized is a callback for the Flow consensus follower. It is called
+// each time a block is finalized by the Flow consensus algorithm.
+func (e *ExecDataSync) OnBlockFinalized(block *model.Block) {
+	blockID := block.BlockID
+	// We push the block ID to the front of the queue; the streamer will try to
+	// download the blocks in a FIFO manner.
+	e.queue.PushFront(blockID)
+
+	e.log.Debug().Hex("block", blockID[:]).Msg("execution record queued for download")
+}
+
+// Next returns the next available block data. It returns an ErrUnavailable if no block
+// data is available at the moment.
+func (e *ExecDataSync) Next() (*entities.BlockExecutionData, error) {
+
+	// If we are not polling already, we want to start polling in the
+	// background. This will try to fill the buffer up until its limit is
+	// reached. It basically means that the cloud streamer will always be
+	// downloading if something is available and the execution tracker is asking
+	// for the next record.
+	go e.poll()
+
+	// If we have nothing in the buffer, we can return the unavailable error,
+	// which will cause the mapper logic to go into a wait state and retry a bit
+	// later.
+	if e.buffer.Len() == 0 {
+		e.log.Debug().Msg("buffer empty, no execution record available")
+		return nil, dps.ErrUnavailable
+	}
+
+	// If we have a record in the buffer, we will just return it. The buffer is
+	// concurrency safe, so there is no problem with popping from the back while
+	// the poll is pushing new items in the front.
+	record := e.buffer.PopBack()
+	return record.(*entities.BlockExecutionData), nil
+}
+
+func (e *ExecDataSync) poll() {
```

Comment:
Rather than creating a new goroutine for each call to poll, what if we used a persistent gorountine and signalled it to run using a channel. This is a common pattern used in `flow-go`. Here's one:
https://github.com/onflow/flow-go/blob/master/engine/collection/ingest/engine.go#L166-L166

This would reduce the overhead from opening many goroutines and reduces the risk of a leak. The [`engine.Notifier`](https://github.com/onflow/flow-go/blob/master/engine/notifier.go) provides a convenient interface for the signalling.